### ISSUE TITLE
Subcomponent fields are not explicitly sorted, leading to Gradle build cache misses

### ISSUE THREAD

**Issue by Egorand** [google/dagger#3169]
Here's a similar issue filed against [InflationInject](https://github.com/cashapp/InflationInject): https://github.com/cashapp/InflationInject/issues/181.

We've got the following `@Subcomponent` declared in our code:

```kotlin
@ActivityScope
@Subcomponent(modules = [MainActivityModule::class])
interface MainActivityComponent : PasscodeView.Injector {
  // other code

  @Subcomponent(modules = [ViewModule::class])
  interface ViewInjector {
    fun inflaterFactory(): InflationInjectFactory
  }
}
```

On CI, Dagger generates the following code for `ViewInjector`:

```kotlin
private static final class ViewInjectorImpl implements MainActivityComponent.ViewInjector {
  // other code

  @SuppressWarnings("unchecked")
  private void initialize() {
    this.refundPaymentPresenterProvider = ...
    this.factoryProvider = ...
    this.reportAbusePresenterProvider = ...
    this.factoryProvider2 = ...
    this.mainTabbedScreensPresenterProvider = ...
    this.inAppNotificationPresenterProvider = ...
    this.factoryProvider3 = ...
    this.activityContactPresenterProvider = ...
    this.factoryProvider4 = ...
    this.receiptDetailsPresenterProvider = ...
    this.factoryProvider5 = ...
    this.receiptPresenterProvider = ...
    this.factoryProvider6 = ...
  }
}
```

And on one of the dev machines the generated code looks like this:

```kotlin
private static final class ViewInjectorImpl implements MainActivityComponent.ViewInjector {
  // other code

  @SuppressWarnings("unchecked")
  private void initialize() {
    this.receiptPresenterProvider = ...
    this.factoryProvider = ...
    this.receiptDetailsPresenterProvider = ...
    this.factoryProvider2 = ...
    this.activityContactPresenterProvider = ...
    this.factoryProvider3 = ...
    this.refundPaymentPresenterProvider = ...
    this.factoryProvider4 = ...
    this.reportAbusePresenterProvider = ...
    this.factoryProvider5 = ...
    this.mainTabbedScreensPresenterProvider = ...
    this.inAppNotificationPresenterProvider = ...
    this.factoryProvider6 = ...
  }
}
```

The difference is that the fields aren't declared in the same order, and even though this doesn't have any effect on code correctness (please correct me if I'm wrong here), Gradle's input comparison will fail leading to a cache miss.

The solution might be to sort entries before generating code, similar to https://github.com/cashapp/InflationInject/pull/182.

**Comment by Chang-Eric** [comment]
I think we might need more information. Are you only seeing this difference in CI vs dev builds? (e.g. are dev builds consistent with other dev builds?) If so, if you have any clues on the difference between your CI and dev build might be good. I didn't look closely at what you are generating, but if your code generator is changing its output that Dagger relies on then that might be the cause of it too.

**Comment by Egorand** [comment]
> Are you only seeing this difference in CI vs dev builds? (e.g. are dev builds consistent with other dev builds?)

TBH I only compared the code generated by CI against the code generated on my local machine.

> If so, if you have any clues on the difference between your CI and dev build might be good.

I suspect I might be using a different JDK vendor than the one our CI uses, which may or may not be an issue. 

I'm not familiar with the implementation details of Dagger, however I dug a bit into [InflationInject](https://github.com/cashapp/InflationInject)'s codegen and found that the contents of generated files are based on the data pulled using `RoundEnvironment.getElementsAnnotatedWith`, which returns a `Set`, so there's no guarantee the entries will appear in a stable order between different JVMs or even different invocations on the same JVM - this I believe was the source of the problem.

**Comment by Chang-Eric** [comment]
Okay, in that case, I'm going to close this issue for now since it sounds like the solution is in InflationInject. Thanks!

**Comment by Egorand** [comment]
Apologies if I wasn't clear, the issue _is_ in Dagger (see attached code) - I just mentioned a similar issue in InflationInject. Asking for this ticket to be reopened as there's a fix needed in Dagger.

**Comment by Chang-Eric** [comment]
Oh sorry, I was reading that incorrectly. I'll reopen the issue, but I think I'd still like more information on whether there is a difference in environment or not that is causing this. Something similar came up in an issue previously (https://github.com/google/dagger/issues/1733#issuecomment-585466988) where we suspected the difference was between the eclipse and javac compilers, and at that time I was not willing to guarantee a stable ordering across different compilers.

If this is an issue of build to build with the same environment that would be something we would be willing to look into, but I think it would be very onerous for us to guarantee it general. Since this could come from a lot of places like anytime we ask a class for anything, its methods, fields, etc, we'd potentially need to sort things and it could turn into quite a lot.

**Comment by Egorand** [comment]
Fair point, and I'm not really sure if the difference in the environment is causing this, and if it's possible to get similar results between re-runs on the same environment - any ideas on how to verify this?

I think Dagger should only explicitly sort entries in cases when they're not naturally ordered in the source code - so hopefully only a subset of all the things Dagger generates (we only see this issue with `@Subcomponent`). An example of natural order here is constructor params - they should appear in generated code in the same order as in the source, and there should be no change in the generated code without a change in the source code. When generated code aggregates entries from multiple source files there's no natural order, and I think it would make sense for Dagger to do the sorting.

I think there's a lot of value in this change to developers who use Gradle caching (most Android developers?) as they would get a lot less cache misses, but I definitely agree with your argument about a possible performance hit of explicit sorting. Maybe make this an opt-in feature?

**Comment by Egorand** [comment]
I've got an update: looks like the original source of the issue was the InflationInject bug that I linked - after merging the fix I don't see any Gradle cache issues in the code that Dagger generates. Given that our `@Subcomponent` depended on the `InflationInjectFactory`, it makes sense that the unstable order of entries in the file was propagated to the subcomponent. I'm gonna close this issue as I believe there's nothing to be done here on the Dagger side, thanks for taking a look!
